### ISSUE TITLE
为什么论文中描述的SPA注意力机制与代码不一致呢

### ISSUE THREAD

**Issue by znqaq** [WhatAShot/ExcelFormer#9]
您好：
观摩您的代码，发现SPA模块与论文中描述不一致。论文中SPA中mask的设置是根据“互信息”的大小来决定的

> ** The function 𝐼(·) represents a measure of feature importance, and
>  we use the “mutual information” metric in this paper. If a feature f𝑖
>  is more informative compared to f𝑗, 𝑀[𝑖, 𝑗] is set −∞ (we use −105
>  in implementation) and thus the (𝑖, 𝑗) grid on the attention map is
>  masked**

但代码中您的mask是这样的：

`        attention_scores = q @ k.transpose(1, 2) / math.sqrt(d_head_key) # b f f
        masks = self.get_attention_mask(attention_scores.shape, attention_scores.device)
        attention = F.softmax(attention_scores + masks, dim=-1)`

`def get_attention_mask(self, input_shape, device):
        bs, _, seq_len = input_shape
        seq_ids = torch.arange(seq_len, device=device)
        attention_mask = seq_ids[None, None, :].repeat(bs, seq_len, 1) <= seq_ids[None, :, None]
        # attention_mask = seq_ids[None, :].repeat(seq_len, 1) <= seq_ids[:, None]
        attention_mask = (1.0 - attention_mask.float()) * -1e4
        return attention_mask`

即直接根据特征输入 的位置来获取mask，与互信息无关
希望能得到您的回答

**Comment by znqaq** [comment]
抱歉，我现在才发现您已经在计算互信息后将特征进行排序了
